src/core/lib/iomgr/error.h

```diff
@@ -0,0 +1,143 @@
+#include <stdint.h>
+
+#include <grpc/support/time.h>
+
+// Opaque representation of an error.
+// Errors are refcounted objects that represent the result of an operation.
+// Ownership laws:
+//  if a grpc_error is returned by a function, the caller owns a ref to that
+//    instance
+//  if a grpc_error is passed to a grpc_closure callback function (functions
+//    with the signature:
+//      void (*f)(grpc_exec_ctx *exec_ctx, void *arg, grpc_error *error))
+//    then those functions do not automatically own a ref to error
+//  if a grpc_error is passed to *ANY OTHER FUNCTION* then that function takes
+//    ownership of the error
+typedef struct grpc_error grpc_error;
+
+typedef enum {
+  GRPC_ERROR_INT_ERRNO,
+  GRPC_ERROR_INT_FILE_LINE,
+  GRPC_ERROR_INT_STATUS_CODE,
+  GRPC_ERROR_INT_WARNING,
+  GRPC_ERROR_INT_STREAM_ID,
+  GRPC_ERROR_INT_GRPC_STATUS,
+  GRPC_ERROR_INT_OFFSET,
+  GRPC_ERROR_INT_INDEX,
+  GRPC_ERROR_INT_SIZE,
+  GRPC_ERROR_INT_HTTP2_ERROR,
+  GRPC_ERROR_INT_TSI_CODE,
+  GRPC_ERROR_INT_SECURITY_STATUS,
+  GRPC_ERROR_INT_WSA_ERROR,
+  GRPC_ERROR_INT_FD,
+} grpc_error_ints;
+
+typedef enum {
+  GRPC_ERROR_STR_DESCRIPTION,
+  GRPC_ERROR_STR_FILE,
+  GRPC_ERROR_STR_OS_ERROR,
+  GRPC_ERROR_STR_SYSCALL,
+  GRPC_ERROR_STR_TARGET_ADDRESS,
+  GRPC_ERROR_STR_GRPC_MESSAGE,
+  GRPC_ERROR_STR_RAW_BYTES,
+  GRPC_ERROR_STR_TSI_ERROR,
+  GRPC_ERROR_STR_FILENAME,
+} grpc_error_strs;
+
+typedef enum {
+  GRPC_ERROR_TIME_CREATED,
+} grpc_error_times;
+
+#define GRPC_ERROR_NONE ((grpc_error *)NULL)
+#define GRPC_ERROR_OOM ((grpc_error *)1)
+#define GRPC_ERROR_CANCELLED ((grpc_error *)2)
+
+const char *grpc_error_string(grpc_error *error);
+void grpc_error_free_string(const char *str);
+
+grpc_error *grpc_error_create(const char *file, int line, const char *desc,
+                              grpc_error **referencing, size_t num_referencing);
+#define GRPC_ERROR_CREATE(desc) \
+  grpc_error_create(__FILE__, __LINE__, desc, NULL, 0)
+
+// Create an error that references some other errors. This function adds a
+// reference to each error in errs - it does not consume an existing reference
+#define GRPC_ERROR_CREATE_REFERENCING(desc, errs, count) \
+  grpc_error_create(__FILE__, __LINE__, desc, errs, count)
+
+//#define GRPC_ERROR_REFCOUNT_DEBUG
+#ifdef GRPC_ERROR_REFCOUNT_DEBUG
+grpc_error *grpc_error_ref(grpc_error *err, const char *file, int line,
+                           const char *func);
+void grpc_error_unref(grpc_error *err, const char *file, int line,
+                      const char *func);
+#define GRPC_ERROR_REF(err) grpc_error_ref(err, __FILE__, __LINE__, __func__)
+#define GRPC_ERROR_UNREF(err) \
+  grpc_error_unref(err, __FILE__, __LINE__, __func__)
+#else
+grpc_error *grpc_error_ref(grpc_error *err);
+void grpc_error_unref(grpc_error *err);
+#define GRPC_ERROR_REF(err) grpc_error_ref(err)
+#define GRPC_ERROR_UNREF(err) grpc_error_unref(err)
+#endif
+
+grpc_error *grpc_error_set_int(grpc_error *src, grpc_error_ints which,
```

Comment:
we need some docs for the role and meaning of error ints, either here or in the `grpc_error_ints` enum